File: Home2WorkAPI/src/main/java/it/gruppoinfor/home2workapi/HomeToWorkClient.java
package it.gruppoinfor.home2workapi;


import com.facebook.stetho.okhttp3.StethoInterceptor;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import it.gruppoinfor.home2workapi.interfaces.LoginCallback;
import it.gruppoinfor.home2workapi.model.Company;
import it.gruppoinfor.home2workapi.model.Guest;
import it.gruppoinfor.home2workapi.model.Location;
import it.gruppoinfor.home2workapi.model.Match;
import it.gruppoinfor.home2workapi.model.Share;
import it.gruppoinfor.home2workapi.model.User;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class HomeToWorkClient {

    public static final String AVATAR_BASE_URL = "http://home2workapi.azurewebsites.net/images/avatar/";
    public static final String COMPANIES_BASE_URL = "http://home2workapi.azurewebsites.net/images/companies/";
    public static final String ACHIEVEMENTS_BASE_URL = "http://home2workapi.azurewebsites.net/images/achievements/";
    private static final String API_BASE_URL = "http://home2workapi.azurewebsites.net/api/";
    private static HomeToWorkClient sInstance;
    private static User sUser;
    private EndpointInterface mAPI;

    private HomeToWorkClient() {
        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .addNetworkInterceptor(new StethoInterceptor())
                .build();

        Gson gson = new GsonBuilder()
                .setDateFormat("yyyy-MM-dd'T'HH:mm:ss")
                .create();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(API_BASE_URL)
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();

        mAPI = retrofit.create(EndpointInterface.class);
    }

    public static void init() {
        sInstance = new HomeToWorkClient();
    }

    public static HomeToWorkClient getInstance() {
        return sInstance;
    }

    /**
     * Ottiene l'utente attualmente autenticato nel client
     *
     * @return User utente autenticato
     */
    public static User getUser() {
        return sUser;
    }

    /**
     * Effettua il login di un utente con le credenziali passate
     *
     * @param email           String Email utente
     * @param password        String Pasword dell'utente
     * @param isPasswordToken boolean Flag per indicare se si sta accedendo con un token (true) oppure con la password (false)
     * @param loginCallback   LoginCallback Callback per il login
     */
    public void login(String email, String password, boolean isPasswordToken, LoginCallback loginCallback) {
        mAPI.login(email, password, isPasswordToken)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(userResponse -> {
                    switch (userResponse.code()) {
                        case 404:
                            loginCallback.onInvalidCredential();
                            break;
                        case 200:
                            sUser = userResponse.body();
                            loginCallback.onLoginSuccess();
                            break;
                        default:
                            loginCallback.onLoginError();
                            break;
                    }
                }, loginCallback::onError);
    }

    /**
     * Imposta nel server il token unico per la FMC Platform
     *
     * @param fcmToken Token generato per Firebase Messaging Cloud platform
     */
    public void setFcmToken(String fcmToken) {
        mAPI.setFCMToken(sUser.getId(), fcmToken)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response -> {
                    // .. nulla
                }, Throwable::printStackTrace);
    }

    /**
     * Carica l'avatar sul server
     *
     * @param avatarBody        Immagine avatar
     * @param onSuccessListener Callback in caso di caricamento avvenuto con successo
     * @param onFailureListener Callback in caso di errori del server
     */
    public void uploadAvatar(MultipartBody.Part avatarBody, OnSuccessListener<ResponseBody> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.uploadAvatar(sUser.getId(), avatarBody)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response -> {
                    if (response.code() == 201)
                        onSuccessListener.onSuccess(response.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code " + response.code()));

                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    /**
     * Ottiene la lista dei match dell'utente
     *
     * @param onSuccessListener Callback in caso di caricamento avvenuto con successo. Restituisce una lista di match
     * @param onFailureListener Callback in caso di errori del server
     */
    public void getUserMatches(OnSuccessListener<List<Match>> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.getMatches(sUser.getId())
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(listResponse -> {
                    if (listResponse.code() == 200)
                        onSuccessListener.onSuccess(listResponse.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code: " + listResponse.code()));
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        onFailureListener.onFailure(new Exception(throwable));
                    }
                });
    }

    /**
     * Modifica le informazioni di un match
     *
     * @param match             Match aggiornato da salvarne in cambiamenti sul server
     * @param onSuccessListener Callback in caso di modifica avvenuta con successo. Restituisce il match aggiornato
     * @param onFailureListener Callback in caso di errori del server
     */
    public void editMatch(Match match, OnSuccessListener<Match> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.editMatch(match)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(matchResponse -> {
                    if (matchResponse.code() == 200)
                        onSuccessListener.onSuccess(matchResponse.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code: " + matchResponse.code()));
                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    /**
     * Modifica le informazioni di un match (Metodo senza callbacks)
     *
     * @param match Match aggiornato da salvarne in cambiamenti sul server
     */
    public void editMatch(Match match) {
        mAPI.editMatch(match)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe();
    }

    /**
     * Crea una nuova condivisione sul server e la avvia
     *
     * @param onSuccessListener Callback in caso di creazione avvenuta con successo. Restituisce la condivisione appena creata
     * @param onFailureListener Callback in caso di errori del server
     */
    public void createShare(OnSuccessListener<Share> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.createShare(sUser.getId())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(shareResponse -> {
                    if (shareResponse.code() == 200)
                        onSuccessListener.onSuccess(shareResponse.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code " + shareResponse.code()));

                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    /**
     * Annulla una condivisione
     *
     * @param share             Condivisione da annullare
     * @param onSuccessListener Callback in caso di eliminazione avvenuta con successo.
     * @param onFailureListener Callback in caso di errori del server.
     */
    public void cancelShare(Share share, OnSuccessListener<ResponseBody> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.cancelShare(share.getId(), sUser.getId())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(shareResponse -> {
                    if (shareResponse.code() == 200)
                        onSuccessListener.onSuccess(shareResponse.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code " + shareResponse.code()));

                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    /**
     * Abbandona una condivisione alla quale si sta partecipando come ospiti.
     *
     * @param share             Condivisione da abbandonare
     * @param onSuccessListener Callback in caso di abbandono avvenuto con successo. Restituisce la condivisione aggiornata.
     * @param onFailureListener Callback in caso di errori del server.
     */
    public void leaveShare(Share share, OnSuccessListener<Share> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.leaveShare(share.getId(), sUser.getId())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(shareResponse -> {
                    if (shareResponse.code() == 200)
                        onSuccessListener.onSuccess(shareResponse.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code " + shareResponse.code()));

                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    /**
     * Espelle l'utente dalla condivisione nella quale si è host
     *
     * @param share             Condivisione
     * @param guest             Utente da espellere
     * @param onSuccessListener Callback in caso di espulsione avvenuta con successo. Restituisce la condivisione aggiornata
     * @param onFailureListener Callback in caso di errori del server.
     */
    public void expelGuest(Share share, Guest guest, OnSuccessListener<Share> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.leaveShare(share.getId(), guest.getUser().getId())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(shareResponse -> {
                    if (shareResponse.code() == 200)
                        onSuccessListener.onSuccess(shareResponse.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code " + shareResponse.code()));

                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    /**
     * Ottiene informazioni su una condivisione con l'ID fornito
     *
     * @param sID               ID della condivisione
     * @param onSuccessListener Callback in caso di caricamento avvenuto con successo. Restituisce la condivisione ottenuta
     * @param onFailureListener Callback in caso di errori del server.
     */
    public void getShare(Long sID, OnSuccessListener<Share> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.getShare(sID)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(shareResponse -> {
                    if (shareResponse.code() == 200)
                        onSuccessListener.onSuccess(shareResponse.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code " + shareResponse.code()));

                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    /**
     * Aggiorna le informazioni dell'utente collegato.
     *
     * @param onSuccessListener Callback in caso di aggiornato avvenuto con successo.
     * @param onFailureListener Callback in caso di errori del server.
     */
    public void updateUser(OnSuccessListener<Void> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.updateUser(sUser)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response -> {
                    if (response.code() == 200) {
                        sUser = response.body();
                        onSuccessListener.onSuccess(null);
                    } else
                        onFailureListener.onFailure(new Exception("Response code " + response.code()));

                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    /**
     * Ottiene la lista delle aziende presenti nel sistema.
     *
     * @param onSuccessListener Callback in caso di caricamento avvenuto con successo. Restituisce la lista delle aziende
     */
    public void getCompanies(OnSuccessListener<List<Company>> onSuccessListener) {
        mAPI.getCompanies()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(companiesResponse -> {
                    if (companiesResponse.code() == 200)
                        onSuccessListener.onSuccess(companiesResponse.body());
                }, Throwable::printStackTrace);
    }

    /**
     * Aggiorna l'utente collegato.
     *
     * @param onSuccessListener Callback in caso di aggiornamento avvenuto con successo
     * @param onFailureListener Callback in caso di errori del server.
     */
    public void refreshUser(OnSuccessListener<Void> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.getUser(sUser.getId())
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(userResponse -> {
                    if (userResponse.code() == 200) {
                        sUser = userResponse.body();
                        onSuccessListener.onSuccess(null);
                    } else
                        onFailureListener.onFailure(new Exception("Response code: " + userResponse.code()));
                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    /**
     * Ottiene la lista delle condivisioni dell'utente, in corso e completate.
     *
     * @param onSuccessListener Callback in caso di caricamento avvenuto con successo. Restituisce la lista delle condivisioni.
     * @param onFailureListener Callback in caso di errori del server.
     */
    public void getUserShares(OnSuccessListener<List<Share>> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.getShares(sUser.getId())
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(listResponse -> {
                    if (listResponse.code() == 200)
                        onSuccessListener.onSuccess(listResponse.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code: " + listResponse.code()));
                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    public void joinShare(Long shareId, android.location.Location joinLocation, OnSuccessListener<Share> onSuccessListener, OnFailureListener onFailureListener) {
        String locationString = joinLocation.getLatitude() + "," + joinLocation.getLongitude();
        mAPI.joinShare(shareId, sUser.getId(), locationString)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(responseBodyResponse -> {
                    if (responseBodyResponse.code() == 200)
                        onSuccessListener.onSuccess(responseBodyResponse.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code: " + responseBodyResponse.code()));
                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    public void completeShare(Share share, android.location.Location joinLocation, OnSuccessListener<Share> onSuccessListener, OnFailureListener onFailureListener) {
        String locationString = joinLocation.getLatitude() + "," + joinLocation.getLongitude();
        mAPI.completeShare(share.getId(), sUser.getId(), locationString)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(responseBodyResponse -> {
                    if (responseBodyResponse.code() == 200)
                        onSuccessListener.onSuccess(responseBodyResponse.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code: " + responseBodyResponse.code()));
                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    public void finishShare(Share share, OnSuccessListener<Share> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.finishShare(share.getId(), sUser.getId())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(responseBodyResponse -> {
                    if (responseBodyResponse.code() == 200)
                        onSuccessListener.onSuccess(responseBodyResponse.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code: " + responseBodyResponse.code()));
                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    public void uploadLocation(long userId, List<Location> locationList, OnSuccessListener<List<Location>> onSuccessListener, OnFailureListener onFailureListener) {
        mAPI.uploadLocations(userId, locationList)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(listResponse -> {
                    if (listResponse.code() == 200)
                        onSuccessListener.onSuccess(listResponse.body());
                    else
                        onFailureListener.onFailure(new Exception("Response code" + listResponse.code()));
                }, throwable -> onFailureListener.onFailure(new Exception(throwable)));
    }

    public Observable<Response<User>> getUser(Long userId) {
        return mAPI.getUser(userId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

}

File: Home2WorkAPI/src/main/java/it/gruppoinfor/home2workapi/HomeToWorkClient.kt
package it.gruppoinfor.home2workapi


import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.OnSuccessListener
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import it.gruppoinfor.home2workapi.auth.AuthCallback
import it.gruppoinfor.home2workapi.auth.AuthService
import it.gruppoinfor.home2workapi.auth.AuthUser
import it.gruppoinfor.home2workapi.chat.Chat
import it.gruppoinfor.home2workapi.chat.ChatService
import it.gruppoinfor.home2workapi.chat.Message
import it.gruppoinfor.home2workapi.company.Company
import it.gruppoinfor.home2workapi.company.CompanyService
import it.gruppoinfor.home2workapi.location.Location
import it.gruppoinfor.home2workapi.match.Match
import it.gruppoinfor.home2workapi.match.MatchService
import it.gruppoinfor.home2workapi.share.Share
import it.gruppoinfor.home2workapi.share.ShareService
import it.gruppoinfor.home2workapi.user.User
import it.gruppoinfor.home2workapi.user.UserProfile
import it.gruppoinfor.home2workapi.user.UserService
import okhttp3.MultipartBody
import okhttp3.ResponseBody

object HomeToWorkClient {

    internal const val BASE_URL_OLD = "http://home2workapi.azurewebsites.net"

    internal const val BASE_URL = "https://hometoworkapi.azurewebsites.net"
    internal const val AVATAR_BASE_URL = "$BASE_URL_OLD/images/avatar/"
    internal const val COMPANIES_BASE_URL = "$BASE_URL/images/companies/"
    internal const val ACHIEVEMENTS_BASE_URL = "$BASE_URL/images/achievements/"
    internal const val HEADER_SESSION_TOKEN = "X-User-Session-Token"

    var user: AuthUser? = null
        private set

    var ongoingShare: Share? = null
        private set

    var inbox: ArrayList<Chat> = ArrayList()

    internal var sessionToken: String = ""

    fun getAuthService(): AuthService {
        return ServiceGenerator.createService(AuthService::class.java)
    }

    fun getChatService(): ChatService {
        return ServiceGenerator.createService(ChatService::class.java)
    }

    fun getCompanyService(): CompanyService {
        return ServiceGenerator.createService(CompanyService::class.java)
    }

    fun getMatchService(): MatchService {
        return ServiceGenerator.createService(MatchService::class.java)
    }

    fun getShareService(): ShareService {
        return ServiceGenerator.createService(ShareService::class.java)
    }

    fun getUserService(): UserService {
        return ServiceGenerator.createService(UserService::class.java)
    }

    /**
     * Effettua il login di un utente con le credenziali passate
     */
    fun login(email: String, password: String, authCallback: AuthCallback) {

        val service = ServiceGenerator.createService(AuthService::class.java)

        service.login(email, password)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->
                    when (response.code()) {
                        404 -> authCallback.onInvalidCredential()
                        200 -> {
                            user = response.body()!!
                            authCallback.onSuccess()
                        }
                        else -> authCallback.onError(null)
                    }
                }, {
                    authCallback.onError(it)
                })

    }

    /**
     * Effettua il login di un utente tramite l'Access Token
     */
    fun login(token: String, authCallback: AuthCallback) {

        val service = ServiceGenerator.createService(AuthService::class.java)

        service.login(token)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->
                    when (response.code()) {
                        404 -> authCallback.onInvalidCredential()
                        200 -> {

                            user = response.body()!!
                            authCallback.onSuccess()

                        }
                        else -> authCallback.onError(null)
                    }
                }, {
                    authCallback.onError(it)
                })

    }

    /**
     * Imposta nel server il token unico per la FMC Platform
     */
    fun updateFcmToken(fcmToken: String?, onSuccessListener: OnSuccessListener<ResponseBody>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(UserService::class.java)

        if (fcmToken == null) return
        service.updateFCMToken(fcmToken)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->
                    if (response.code() == 200)
                        onSuccessListener.onSuccess(response.body())
                    else
                        onFailureListener.onFailure(Exception("Response code " + response.code()))
                }, { it.printStackTrace() })

    }

    /**
     * Carica l'avatar sul server
     */
    fun uploadAvatar(avatarBody: MultipartBody.Part, onSuccessListener: OnSuccessListener<ResponseBody>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(UserService::class.java)

        service.uploadAvatar(avatarBody)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->

                    if (response.code() == 201)
                        onSuccessListener.onSuccess(response.body())
                    else
                        onFailureListener.onFailure(Exception("Response code " + response.code()))

                }, { it.printStackTrace() })

    }

    /**
     * Aggiorna le informazioni dell'utente collegato.
     */
    fun updateUser(onSuccessListener: OnSuccessListener<Void>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(UserService::class.java)

        service.edit(user!!)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->

                    onSuccessListener.onSuccess(null)

                }, { it.printStackTrace() })

    }

    /**
     * Aggiorna l'utente collegato.
     */
    fun refreshUser(onSuccessListener: OnSuccessListener<Void>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(UserService::class.java)

        service.get()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->

                    if (response.code() == 200) {
                        user = response.body()!!
                        onSuccessListener.onSuccess(null)
                    } else
                        onFailureListener.onFailure(Exception("Response code: " + response.code()))

                }, { it.printStackTrace() })

    }

    /**
     * Ottiene il profilo dell'utente collegato
     */
    fun getProfile(onSuccessListener: OnSuccessListener<UserProfile>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(UserService::class.java)

        service.getProfile()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ userResponse ->

                    onSuccessListener.onSuccess(userResponse)

                }, { it.printStackTrace() })

    }

    /**
     * Ottiene la lista dei match dell'utente
     */
    fun getMatchList(onSuccessListener: OnSuccessListener<List<Match>>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(UserService::class.java)

        service.getMatchList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ listResponse ->

                    onSuccessListener.onSuccess(listResponse)

                }, { throwable ->
                    onFailureListener.onFailure(Exception(throwable))
                })

    }

    /**
     * Modifica le informazioni di un match
     */
    fun editMatch(match: Match, onSuccessListener: OnSuccessListener<Match>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(MatchService::class.java)

        service.editMatch(match)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ matchResponse ->

                    onSuccessListener.onSuccess(matchResponse)

                }, { it.printStackTrace() })
    }

    /**
     * Modifica le informazioni di un match (Metodo senza callbacks)
     */
    fun editMatch(match: Match) {

        val service = ServiceGenerator.createService(MatchService::class.java)

        service.editMatch(match)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe()

    }

    /**
     * Ottiene la condivisione in corso
     */
    fun getOngoingShare(onSuccessListener: OnSuccessListener<Share>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(ShareService::class.java)

        service.getOngoingShare()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ shareResponse ->

                    if (shareResponse.code() == 200) {

                        ongoingShare = shareResponse.body()
                        onSuccessListener.onSuccess(shareResponse.body())

                    } else
                        onFailureListener.onFailure(Exception("Response code " + shareResponse.code()))

                }, { it.printStackTrace() })

    }

    /**
     * Crea una nuova condivisione sul server e la avvia
     */
    fun createNewShare(onSuccessListener: OnSuccessListener<Share>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(ShareService::class.java)

        service.createNewShare(user!!.id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ shareResponse ->

                    if (shareResponse.code() == 200) {

                        ongoingShare = shareResponse.body()
                        onSuccessListener.onSuccess(shareResponse.body())

                    } else
                        onFailureListener.onFailure(Exception("Response code " + shareResponse.code()))

                }, { it.printStackTrace() })

    }

    /**
     * Annulla una condivisione
     */
    fun cancelShare(shareId: Long, onSuccessListener: OnSuccessListener<ResponseBody>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(ShareService::class.java)

        service.cancelShare(shareId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ shareResponse ->

                    if (shareResponse.code() == 200) {
                        ongoingShare = null
                        onSuccessListener.onSuccess(shareResponse.body())
                    } else
                        onFailureListener.onFailure(Exception("Response code " + shareResponse.code()))

                }, { it.printStackTrace() })

    }

    /**
     * Abbandona una condivisione alla quale si sta partecipando come ospiti.
     */
    fun leaveShare(share: Share, onSuccessListener: OnSuccessListener<Share>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(ShareService::class.java)

        service.leaveShare(share.id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ shareResponse ->

                    if (shareResponse.code() == 200) {
                        ongoingShare = null
                        onSuccessListener.onSuccess(shareResponse.body())
                    } else
                        onFailureListener.onFailure(Exception("Response code " + shareResponse.code()))

                }, { it.printStackTrace() })

    }

    /**
     * Espelle l'utente dalla condivisione nella quale si è host
     */
    fun banGuestFromShare(shareId: Long?, guestId: Long?, onSuccessListener: OnSuccessListener<Share>, onFailureListener: OnFailureListener) {

        if (shareId == null || guestId == null)
            onFailureListener.onFailure(IllegalArgumentException())
        else {
            val service = ServiceGenerator.createService(ShareService::class.java)

            service.banGuestFromShare(shareId, guestId)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ shareResponse ->

                        if (shareResponse.code() == 200)
                            onSuccessListener.onSuccess(shareResponse.body())
                        else
                            onFailureListener.onFailure(Exception("Response code " + shareResponse.code()))

                    }, { it.printStackTrace() })
        }

    }

    /**
     * Ottiene informazioni su una condivisione con l'ID fornito
     */
    fun getShare(shareId: Long?, onSuccessListener: OnSuccessListener<Share>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(ShareService::class.java)

        service.getShareById(shareId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ shareResponse ->

                    if (shareResponse.code() == 200)
                        onSuccessListener.onSuccess(shareResponse.body())
                    else
                        onFailureListener.onFailure(Exception("Response code " + shareResponse.code()))

                }, { it.printStackTrace() })

    }

    /**
     * Ottiene la lista delle aziende presenti nel sistema.
     */
    fun getCompanyList(onSuccessListener: OnSuccessListener<ArrayList<Company>>) {

        val service = ServiceGenerator.createService(CompanyService::class.java)

        service.companies
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ companiesResponse ->

                    if (companiesResponse.code() == 200)
                        onSuccessListener.onSuccess(companiesResponse.body())

                }, { it.printStackTrace() })

    }

    /**
     * Ottiene la lista delle condivisioni dell'utente, in corso e completate.
     */
    fun getShareList(onSuccessListener: OnSuccessListener<List<Share>>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(UserService::class.java)

        service.getShareList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ listResponse ->

                    onSuccessListener.onSuccess(listResponse)

                }, { it.printStackTrace() })
    }

    /**
     * Unisce ad una condivisione l'utente collegato
     */
    fun joinShare(shareId: Long?, joinLocation: android.location.Location, onSuccessListener: OnSuccessListener<Share>, onFailureListener: OnFailureListener) {

        if (shareId == null)
            onFailureListener.onFailure(IllegalArgumentException())
        else {
            val locationString = joinLocation.latitude.toString() + "," + joinLocation.longitude

            val service = ServiceGenerator.createService(ShareService::class.java)

            service.joinShare(shareId, locationString)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ responseBodyResponse ->

                        if (responseBodyResponse.code() == 200)
                            onSuccessListener.onSuccess(responseBodyResponse.body())
                        else
                            onFailureListener.onFailure(Exception("Response code: " + responseBodyResponse.code()))

                    }, { it.printStackTrace() })
        }

    }

    /**
     * Imposta come completata una condivisione
     */
    fun completeShare(share: Share, joinLocation: android.location.Location, onSuccessListener: OnSuccessListener<Share>, onFailureListener: OnFailureListener) {

        val locationString = joinLocation.latitude.toString() + "," + joinLocation.longitude

        val service = ServiceGenerator.createService(ShareService::class.java)

        service.completeShare(share.id, locationString)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ responseBodyResponse ->

                    if (responseBodyResponse.code() == 200)
                        onSuccessListener.onSuccess(responseBodyResponse.body())
                    else
                        onFailureListener.onFailure(Exception("Response code: " + responseBodyResponse.code()))

                }, { it.printStackTrace() })

    }

    /**
     * Completa una condivsione da parte di un ospite
     */
    fun finishShare(share: Share, onSuccessListener: OnSuccessListener<Share>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(ShareService::class.java)

        service.finishShare(share.id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ responseBodyResponse ->

                    if (responseBodyResponse.code() == 200)
                        onSuccessListener.onSuccess(responseBodyResponse.body())
                    else
                        onFailureListener.onFailure(Exception("Response code: " + responseBodyResponse.code()))

                }, { it.printStackTrace() })

    }

    /**
     * Effettua la sincronizzazione delle posizioni dell'utente
     */
    fun uploadLocations(locationList: List<Location>, onSuccessListener: OnSuccessListener<List<Location>>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(UserService::class.java)

        service.uploadLocations(locationList)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ listResponse ->

                    if (listResponse.code() == 200)
                        onSuccessListener.onSuccess(listResponse.body())
                    else
                        onFailureListener.onFailure(Exception("Response code" + listResponse.code()))

                }, { it.printStackTrace() })

    }

    /**
     * Ottiene l'utente con l'id fornito
     */
    fun getUserById(userId: Long?, onSuccessListener: OnSuccessListener<User>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(UserService::class.java)

        service.getUserById(userId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->

                    if (response.code() == 200) {
                        onSuccessListener.onSuccess(response.body())
                    } else
                        onFailureListener.onFailure(Exception("Response code: " + response.code()))

                }, { it.printStackTrace() })

    }

    /**
     * Ottiene il profilo dell'utente con l'id fornito
     */
    fun getUserProfileById(userId: Long?, onSuccessListener: OnSuccessListener<UserProfile>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(UserService::class.java)

        service.getUserProfileById(userId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ userResponse ->

                    onSuccessListener.onSuccess(userResponse)

                }, { it.printStackTrace() })

    }

    /**
     * Ottiene la lista delel chat dell'utente collegato
     */
    fun getChatList(onSuccessListener: OnSuccessListener<List<Chat>>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(UserService::class.java)

        service.getChatList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .onErrorReturn {
                    onFailureListener.onFailure(Exception(it))
                    emptyList()
                }
                .subscribe({ response ->

                    onSuccessListener.onSuccess(response)

                }, { it.printStackTrace() })

    }

    /**
     * Ottiene i messaggi della chat con l'id fornito
     */
    fun getChatMessageList(chatId: Long, onSuccessListener: OnSuccessListener<List<Message>>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(ChatService::class.java)

        service.getChatMessageList(chatId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->

                    onSuccessListener.onSuccess(response)

                }, { it.printStackTrace() })

    }

    /**
     * Invia un messaggio alla chat con l'id fornito
     */
    fun sendMessageToChat(chatId: Long, text: String, onSuccessListener: OnSuccessListener<ResponseBody>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(ChatService::class.java)

        service.sendMessageToChat(chatId, text)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->

                    onSuccessListener.onSuccess(response)

                }, { it.printStackTrace() })

    }

    fun newChat(recipientId: Long, onSuccessListener: OnSuccessListener<Chat>, onFailureListener: OnFailureListener) {

        val service = ServiceGenerator.createService(ChatService::class.java)

        service.newChat(recipientId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->

                    onSuccessListener.onSuccess(response)

                }, {
                    it.printStackTrace()
                })

    }


}
